h/models/group.py

```diff
@@ -80,6 +80,8 @@ class Group(Base, mixins.Timestamps):
         'User', secondary='user_group', backref=sa.orm.backref(
             'groups', order_by='Group.name'))
 
+    scopes = sa.orm.relationship('GroupScope', backref='group', cascade='all, delete-orphan')
```

Comment:
The `cascade` argument here is what ensures that if a group is deleted, or a group's scopes are changed so that some groupscopes are no longer referenced, the groupscopes automatically get deleted.

Reply:
(I think this is just sqlalchemy stuff though, it doesn't produce an `ON DELETE CASCADE` in SQL, so deleting a group in SQL may not delete groupscopes, will look into this.)

Reply:
> I think this is just sqlalchemy stuff though, it doesn't produce an ON DELETE CASCADE in SQL, so deleting a group in SQL may not delete groupscopes, will look into this

I would be interested to learn how this works, please post your findings.

Reply:
Ok, I've added `nullable=False` to the `group_id` column of `groupscope` to prevent orphan `groupscope` rows. And I've added `ondelete='cascade'` to the `ForeignKey` in `GroupScope` to enable convenient deletion of groups when working in SQL directly. Details:

The `cascade` argument to `relationship()` is documented here: <http://docs.sqlalchemy.org/en/rel_1_1/orm/cascades.html>.

Normally, without `cascade='all, delete-orphan'` on the relationship, you'd have two problems:

1. Deleting a group (`db_session.delete(group)`) causes an `IntegrityError` from Postgres because there are rows in the `groupscope` table with foreign keys that reference the deleted group. You would have to manually delete the groupscopes first.

2. Removing the scopes from a group (`group.scopes = []`) also causes an `IntegrityError` from Postgres because sqlalchemy tries to set the `group_id` column of these `groupscope` rows to `NULL` which violates the `NOT NULL` constraint on `groupscope.group_id`. (Or, if there were no such `NOT NULL` constraint, then you wouldn't get an error but you would get orphan `groupscope` rows with `NULL` `group_id`s left in the DB.) Again, you would have to manually delete the groupscopes first.

With the `cascade='all, delete-orphan'` though, sqlalchemy changes its behaviour so that either of the above would automatically delete the `groupscope` rows for you.

This `cascade` is a sqlalchemy feature (implemented by sqlalchemy keeping track of things itself and emitting additional SQL `DELETE` commands), it makes no difference to the table definition that ends up in Postgres.

If you tried to delete a group directly in SQL you would get an error because you need to delete the groupscopes that're referencing it first. I've now added an `ondelete='cascade'` argument to the `ForeignKey` in the `GroupScope` table though. _That_ does mean that `ON DELETE CASCADE` gets added to the `FOREIGN KEY` constraint on the `groupscope` table in the SQL, so now you _can_ just delete a group in SQL and the corresponding `groupscopes` are automatically deleted.

Because `groupscope.group_id` is a non-nullable foreign key, in all cases sqlalchemy or sql, even without either of the cascade arguments, it's never possible to have orphan `groupscope` rows in the DB. What the cascades do is automatically delete groupscopes instead of crashing, if you just try to delete a group or edit group.scopes without deleting the scopes yourself first.

Reply:
I follow. Thanks for the explanation 🙂